### DISCUSSION TITLE
Before using show_pdf_page, clean_page resulted in loss of content.

### DISCUSSION THREAD

**Discussion by izerui** [pymupdf/PyMuPDF#3359] [Looking for help] [answered]
Rearranged the problem and wrote the test case:.
Question.
1. if use `clean_page` will lead to the third pdf page content loss
2. if the use of `wrap_page` will lead to (1) pdf, (2) pdf, (4) pdf annotation coordinate error
3. if you do not operate the page, in addition to (3) pdf, are displayed or annotated coordinates of the problem

test code:
```
import os

import fitz


def bake_document(doc):
    source_file_pdf = fitz.mupdf.pdf_document_from_fz_document(doc)
    fitz.mupdf.pdf_bake_document(source_file_pdf, 1, 1)
    pass


def handle_page(page, type):
    if type == 0:
        page.insert_text(point=fitz.Point(300, 300), text='no_operation', fontsize=48, color=[1, 0, 0])
    elif type == 1:
        page.clean_contents()
        page.insert_text(point=fitz.Point(300, 300), text='clean_contents', fontsize=48, color=[1, 0, 0])
    elif type == 2:
        page.wrap_contents()
        page.insert_text(point=fitz.Point(300, 300), text='wrap_contents', fontsize=48, color=[1, 0, 0])
    pass


def test():
    docs = [
        fitz.open('28205N61101AAA (1).pdf'),
        fitz.open('mt_03_22318er.pdf'),
        fitz.open('401-020605-00.pdf'),
        fitz.open('mt_04_23024cc.pdf')
    ]

    new_doc = fitz.open()
    for doc in docs:
        new_doc.insert_pdf(docsrc=doc)

        page = doc[0]
        print(doc.xref_object(page.xref))

        # TODO Switch from 0 to 2 to reproduce the problem
        # handle_page(page, 0) # no operation  error: 1,2,4
        # handle_page(page, 1) # clean_contents error: 3
        handle_page(page, 2) # wrap_contents error: 1,2,4

        bake_document(doc)

        new_page = new_doc.new_page(width=page.cropbox.width, height=page.cropbox.height)
        new_page.show_pdf_page(new_page.cropbox, doc, 0)
        doc.close()
    new_doc.ez_save('x.pdf')
    import webbrowser
    browser = webbrowser.get('chrome')
    browser.open(f'file:///{os.path.realpath("x.pdf")}')
```

[mt_03_22318er.pdf](https://github.com/pymupdf/PyMuPDF/files/14986614/mt_03_22318er.pdf)
[401-020605-00.pdf](https://github.com/pymupdf/PyMuPDF/files/14986617/401-020605-00.pdf)
[mt_04_23024cc.pdf](https://github.com/pymupdf/PyMuPDF/files/14986618/mt_04_23024cc.pdf)
[28205N61101AAA (1).pdf](https://github.com/pymupdf/PyMuPDF/files/14986619/28205N61101AAA.1.pdf)

a whole test case: `handle_page` used to switch between different test effects.

Can you help me, this problem has been bothering me for a long time, have not been able to find the most appropriate way to ensure that all pdf can be displayed properly page and annotation coordinates are also normal!

**Comment by JorjMcKie** [comment]
Did you communicate with @robinwatts about this already on Discord?
As a circumvention, try `page.wrap_contents()` instead of `page.clean_contents()`. This method does not touch the existing `/Contents` but just wraps it with push/pop graphic state commands.

**Reply by izerui** [reply]
Looking forward to your reply and @robinwatts , you said wrap_contents I tried, this will lead to the first pdf file to picture display location error.  If I use clean_contents , then the second pdf file will appear content is lost and the picture does not display, if both do not use, then the first pdf file is not only content is lost, and the picture does not display!

Since I'm using an internal function, I don't know what the processing logic for each of clean_contents and wrap_contents looks like, and if I can avoid it.

Thank you for your interest.

**Comment by JorjMcKie** [comment]
Found the problem!
The following script works:
```python
import fitz


def test():
    """
    clean_contents 后导致内容丢失，但是通过mac的 preview.app 可以查看
    """
    files = ["mt_03_22318er_0_806.pdf", "丢失大量内容401-020605-00.pdf"]

    rect = [0, 0, 200, 300]

    target_doc = fitz.open()
    for file in files:
        doc = fitz.open(file)
        for page in doc:
            page.clean_contents(sanitize=0)
            page.insert_image(rect, filename="nur-ruhig.jpg")
            page.draw_rect(rect, color=(1, 0, 0))
            new_page = target_doc.new_page(width=842, height=595)
            new_page.show_pdf_page(new_page.rect, doc, page.number)
        doc.close()
    target_doc.save("x.pdf")
    target_doc.close()
```

**Reply by izerui** [reply]
Thank you for your reply, but I provided by you through the sample code, after the implementation, or the same problem, open the generated pdf file, the second page or display errors, part of the content is not visible.

**Comment by izerui** [comment]
Almighty @JorjMcKie , please help me! There's nothing I can do.

**Reply by JorjMcKie** [reply, marked as answer]
You have to wait for the new version which fixes a number of things.
Most importantly, you will no longer have to use `clean_contents()`.

**Reply by izerui** [reply]
yes sir, i got it! I'm back waiting for the new version to use it first time . haha
![WX20240416-204210@2x](https://github.com/pymupdf/PyMuPDF/assets/1773233/b763f382-4b59-445d-8c27-0ca935cc907b)

**Reply by JorjMcKie** [reply]
Your problem should be solved now with v1.24.2.

**Reply by izerui** [reply]
Yes, I've used it through the new version and it really doesn't need to use `clean_contents` or `wrap_contents` anymore, thank you very much!!!!
